rs/p2p/artifact_manager/src/lib.rs

```diff
@@ -169,52 +162,29 @@ fn run_artifact_processor<
     Box::new(ArtifactProcessorJoinGuard::new(handle, shutdown))
 }
 
-enum StreamState<T> {
-    Value(T),
-    NoNewValueAvailable,
-    EndOfStream,
-}
-
-async fn read_batch<T, S: Stream<Item = T> + Send + Unpin + 'static>(
-    mut stream: Pin<&mut S>,
-    recv_timeout: Duration,
-) -> Option<Vec<T>> {
-    let mut stream = std::pin::Pin::new(&mut stream);
-    match timeout(recv_timeout, stream.next()).await {
-        Ok(Some(first_value)) => {
-            let mut res = vec![first_value];
-            // We ignore the end of stream and empty value states.
-            while let StreamState::Value(value) =
-                std::future::poll_fn(|cx| match stream.as_mut().poll_next(cx) {
-                    Poll::Pending => Poll::Ready(StreamState::NoNewValueAvailable),
-                    Poll::Ready(Some(v)) => Poll::Ready(StreamState::Value(v)),
-                    // Stream has finished because the abortable broadcast/p2p has stopped.
-                    // This is infallible.
-                    Poll::Ready(None) => Poll::Ready(StreamState::EndOfStream),
-                })
-                .await
-            {
-                res.push(value)
-            }
-            Some(res)
-        }
+async fn read_batch<T>(receiver: &mut Receiver<T>, recv_timeout: Duration) -> Option<Vec<T>> {
+    let mut artifacts = vec![];
+    match timeout(
+        recv_timeout,
+        receiver.recv_many(&mut artifacts, MAX_P2P_IO_CHANNEL_SIZE),
+    )
+    .await
+    {
+        Ok(received) if received > 0 => Some(artifacts),
```

Comment:
I think something like this should work? Maybe it's more readable
```suggestion
        Ok(1..)  => Some(artifacts),
        Ok(0) => None,
```

Reply:
That's great, thanks!